File: Posda/c_source/ContourToBitmap.c
/*
#$Source: /home/bbennett/pass/archive/Posda/c_source/ContourToBitmap.c,v $
#$Date: 2014/05/14 15:43:21 $
#$Revision: 1.13 $
#
#Copyright 2008, Bill Bennett
# Part of the Posda package
# Posda may be copied only under the terms of either the Artistic License or the
# GNU General Public License, which may be found in the Posda Distribution,
# or at http://posda.com/License.html
#

# 
# All of the parameters on the command line are of the form:
#  <name>=<value> (no spaces)
# parameter order does not matter.
# Here are the possible parameters:
#  in=<number of first input fd>
#  out=<number of output fd>
#  status=<number of status fd>
#  rows=<number of rows in bitmap output>
#  cols=<number of cols in bitmap output>
#  ulx=<x coordinate of upper left point>
#  uly=<y coordinate of upper left point>
#  ulz=<z coordinate of upper left point>
#  rowspc=<spacing between columns>
#  colspc=<spacing between rows>
#
# On the input socket, the format of the contours is the following:
# BEGIN CONTOUR             - marks the beginning of a contour
# row1,col1                     - first point in contour
# row2,col2                     - next point
# ...                           - and so on
#                               -   not necessarily integer
#                               -   may be negative
# rown,coln                     - last point
# END CONTOUR
# ...                       - repeat for as many contours as you have
#
*/
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdarg.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

char *prog_name;
int input_fh;
int output_fh;
int status_fh;
int rows;
int cols;
int debug = 0;
float ulx;
float uly;
float ulz;
float rowspc;
float colspc;
char *error_message;
FILE *contour_file;
FILE *output_fp;
char *contour_buffer;

typedef struct point{
  float x;
  float y;
} point;
typedef struct point_list{
  point this;
  void *next;
} point_list;

typedef struct line{
  point from;
  point to;
} line;

point_list *point_list_first = NULL;
point_list *point_list_last = NULL;
int point_list_count;

point_list *contours[1024 * 4]; /* only allows 1024 * 4 contours per slice */
int contour_length[1024 * 4];
int contour_i;
line *all_lines;
int line_i;
float intersections[65536 * 4]; /* only allows 65535 * 4 intersections per row */
int int_i, c_i;
/* loop variables */
float x, y, x_1, x_2, y_1, y_2;
int o_polarity, bit_count, count;
int l_polarity;
char output_data;

int total_ones_sent = 0;
int total_zeros_sent = 0;
int total_bits_sent = 0;
int loop_count = 0;

/* Define debug_print to be no op for production */
#define debug_print(args...)
void debug_print_unused(const char* format, ...){
  va_list argptr;
  va_start(argptr, format);
  if (! debug) return;
  fprintf(stderr, "%s (%f): ", prog_name, ulz);
  vfprintf(stderr, format, argptr);
  va_end(argptr);
}

int cmpfloat(const void *a, const void *b){
  if(*(float *)a == *(float *)b)return 0;
  if(*(float *)a < *(float *)b) return -1;
  return 1;
}
int main(int argc, char *argv[]){
  int i, j, ii;
  char *v, *k;
  int scan_mode;
  point_list *p;
  float n;
  prog_name = argv[0];
  for (i = 1; i < argc; i++){
    k = strtok(argv[i], "=");
    v = strtok(NULL, "\n");
    if(strcmp(k, "in") == 0){
      input_fh = atoi(v);
    } else if (strcmp(k, "out") == 0){
      output_fh = atoi(v);
    } else if (strcmp(k, "status") == 0){
      status_fh = atoi(v);
    } else if (strcmp(k, "rows") == 0){
      rows = atoi(v);
    } else if (strcmp(k, "cols") == 0){
      cols = atoi(v);
    } else if (strcmp(k, "ulx") == 0){
      ulx = atof(v);
    } else if (strcmp(k, "uly") == 0){
      uly = atof(v);
    } else if (strcmp(k, "ulz") == 0){
      ulz = atof(v);
    } else if (strcmp(k, "rowspc") == 0){
      rowspc = atof(v);
    } else if (strcmp(k, "colspc") == 0){
      colspc = atof(v);
    } else if (strcmp(k, "debug") == 0){
      debug = atoi(v);
    } else {
      fprintf(stderr, "%s: unknown param: %s\n", prog_name, k);
      exit(-1);
    }
  }

  debug_print("parsing of params is complete\n", prog_name);
  debug_print("%s = %d\n", "in", input_fh);
  debug_print("%s = %d\n", "out", output_fh);
  debug_print("%s = %d\n", "status", status_fh);
  debug_print("%s = %d\n", "rows", rows);
  debug_print("%s = %d\n", "cols", cols);
  debug_print("%s = %f\n", "ulx", ulx); 
  debug_print("%s = %f\n", "uly", uly);
  debug_print("%s = %f\n", "ulz", ulz);
  debug_print("%s = %f\n", "rowspc", rowspc);
  debug_print("%s = %f\n", "colspc", colspc);
  debug_print("%s = %d\n", "debug", debug);

  contour_i = 0;
  contour_buffer = calloc(65536 * 4, 1); /* Parent needs to insure this is 
                                       long enough */
  contour_file = fdopen(input_fh, "r");
  if(contour_file == NULL){
    i = asprintf(&error_message, 
      "%s: can't fdopen input_fh (%d)", prog_name, input_fh);
    perror(error_message);
    exit(-1);
  }
  output_fp = fdopen(output_fh, "w");
  if(output_fp == NULL){
    i = asprintf(&error_message, 
      "%s: can't fdopen output_fh (%d)", prog_name, output_fh);
    perror(error_message);
    exit(-1);
  }
  scan_mode = 0; /* scanning for a contour */
  debug_print("Start scanning\n");
  while(1){
    v = fgets(contour_buffer, 65536 * 4, contour_file);
    if(v != NULL){
      debug_print("Data read, len: %d\n", strlen(v));
    } else {
      debug_print("At EOF\n");
    }
    if(v == NULL) break;
    if(scan_mode == 0 ){
      if(strncmp(v, "BEGIN CONTOUR", 13) != 0) continue;
      debug_print("BEGIN CONTOUR\n");
      scan_mode = 1; /* in countour */
      point_list_first = NULL;
      point_list_last = NULL;
      point_list_count = 0;
      continue;
    }
    if(strncmp(v, "END CONTOUR", 11) == 0){
      /* here's where you close the contour and add it to the list*/
      debug_print("END CONTOUR\n");
      if(
        point_list_first->this.x != point_list_last->this.x ||
        point_list_first->this.y != point_list_last->this.y
      ){
        p = calloc(sizeof(point_list), 1);
        p->this.x = point_list_first->this.x;
        p->this.y = point_list_first->this.y;
        point_list_last->next = p;
        point_list_last = p;
        point_list_count += 1;
      }
      /* push this contour onto the list and look for the next */
      contours[contour_i] = point_list_first;
      contour_length[contour_i] = point_list_count;
      point_list_first = NULL;
      point_list_last = NULL;
      point_list_count = 0;
      contour_i += 1;
      scan_mode = 0;
      continue;
    }
    debug_print("Data read, len: %d\n", strlen(v));
    v = strtok(v, "\\");
    while(v != NULL){
      k = strtok(NULL, "\\");
      if(k == NULL){
        fprintf(stderr, 
          "%s: odd number of numbers in contour", prog_name);
        exit(-1);
      }
      p = calloc(sizeof(point_list), 1);
      p->this.x = atof(v);
      p->this.y = atof(k);
      debug_print("Point: %f,%f\n", p->this.x, p->this.y);

      if(point_list_first == NULL){
        point_list_first = p;
        point_list_last = p;
      } else if (point_list_last == NULL){
        i = asprintf(&error_message, 
          "%s: first is not null and last is null", prog_name);
        perror(error_message);
        exit(-1);
      } else {
        point_list_last->next = p;
        point_list_last = p;
      }
      point_list_count += 1;
      v = strtok(NULL, "\\");
    }
  }
  debug_print("End scanning\n");
  /* here we count up the number of line segments total in all contours */
  for (i = 0, j = 0; i < contour_i; i++){
    j += contour_length[i] - 1;
  }
  /* Now make a list of all line segments */
  all_lines = calloc(sizeof(line), j);
  for(i = 0, line_i = 0; i < j; i++){
    point_list_first = contours[i];
    for(j = 0; j < contour_length[i] - 1; j++, line_i++){
      all_lines[line_i].from.x = point_list_first->this.x;
      all_lines[line_i].from.y = point_list_first->this.y;
      point_list_first = point_list_first->next;
      if(point_list_first == NULL){
        fprintf(stderr, 
          "%s odd number of points in contour when collecting segments\n",
          prog_name);
        exit(-1);
      }
      all_lines[line_i].to.x = point_list_first->this.x;
      all_lines[line_i].to.y = point_list_first->this.y;
    }
  }
  /* Outer loop starts here */
  for(
    i = 0, o_polarity = 0, bit_count = 0, y = uly;
    i < rows;
    i++, y += rowspc
  ){
    /* build array of all crossing points for this row */
    for(
      ii = 0, int_i = 0;
      ii < line_i;
      ii++ 
    ){
      if(all_lines[ii].from.y < all_lines[ii].to.y){
        x_1 = all_lines[ii].from.x;
        y_1 = all_lines[ii].from.y;
        x_2 = all_lines[ii].to.x;
        y_2 = all_lines[ii].to.y;
      } else {
        x_2 = all_lines[ii].from.x;
        y_2 = all_lines[ii].from.y;
        x_1 = all_lines[ii].to.x;
        y_1 = all_lines[ii].to.y;
      }
      if(y_1 < y && y_2 >= y){
        /* calculate and add intersection */
        intersections[int_i] = x_1 + (x_2 - x_1) * ((y - y_1)/(y_2 - y_1));
        int_i += 1;
      } else {

      }
    }
    /* for debug, see that we have an even number of intersections */
    if((int_i & 1) == 1){
      fprintf(stderr, "%s odd number of intersections\n", prog_name);
    }
    /* sort the intersections */
    qsort(intersections, int_i, sizeof(float), cmpfloat);
    /* Inner loop starts here */
    for(
      j = 0, x = ulx, c_i = 0, l_polarity = 0;
      j < cols;
      j++, x += colspc
    ){
      loop_count += 1;
      while(intersections[c_i] < x && c_i < int_i){
        l_polarity = !l_polarity;
        c_i += 1;
      }
      if(l_polarity != o_polarity){
        while(bit_count > 0){
          if(bit_count > 127){
            bit_count -= 127;
            count = 127;
          } else {
            count = bit_count;
            bit_count = 0;
          }
          if(o_polarity){
            output_data = 0x80 + count;
          } else {
            output_data = count;
          }
          if (fputc(output_data, output_fp) == EOF) {
            i = asprintf(&error_message, 
              "%s: Error writing binned data", prog_name);
            perror(error_message);
            exit(-1);
          }
        }
      }
      o_polarity = l_polarity;
      bit_count += 1;
      if(bit_count > 127){
        bit_count = 1;
        count = 127;
        if(o_polarity){
          output_data = 0x80 + count;
        } else {
          output_data = count;
        }
        if (fputc(output_data, output_fp) == EOF) {
          i = asprintf(&error_message, 
            "%s: Error writing binned data", prog_name);
          perror(error_message);
          exit(-1);
        }
      }
    }
  }
  if(bit_count){
        while(bit_count > 0){
          if(bit_count > 127){
            bit_count -= 127;
            count = 127;
          } else {
            count = bit_count;
            bit_count = 0;
          }
          if(o_polarity){
            output_data = 0x80 + count;
          } else {
            output_data = count;
          }
          if (fputc(output_data, output_fp) == EOF) {
            i = asprintf(&error_message, 
              "%s: Error writing binned data", prog_name);
            perror(error_message);
            exit(-1);
          }
        }
  }
  fclose(contour_file);
  fclose(output_fp);
  /* inform parent we're done */
  i = asprintf(&v, "OK\n");
  write(status_fh, (void *) v, i);
  free(v);
}

File: Posda/bin/test/ContourToBitMap.pl
#!/usr/bin/perl -w
#$Source: /home/bbennett/pass/archive/Posda/bin/test/ContourToBitMap.pl,v $
#$Date: 2011/09/22 13:59:59 $
#$Revision: 1.9 $
#
#Copyright 2011, Bill Bennett
# Part of the Posda package
# Posda may be copied only under the terms of either the Artistic License or the
# GNU General Public License, which may be found in the Posda Distribution,
# or at http://posda.com/License.html
#

# Streamed Pixel operations
# This program accepts contours on an fd and writes a bitmap
# on another fd of the points contained within the contours
# The fd's have been opened by the parent process...

# What to do is specified on the command line.
# All of the parameters on the command line are of the form:
#  <name>=<value> (no spaces)
# parameter order does not matter.
# Here are the possible parameters:
#  in=<number of first input fd>
#  out=<number of output fd>
#  status=<number of status fd>
#  rows=<number of rows in bitmap output>
#  cols=<number of cols in bitmap output>
#  ulx=<x coordinate of upper left point>
#  uly=<y coordinate of upper left point>
#  ulz=<z coordinate of upper left point>
#  rowdcosx=<x of row direction cosine>
#  rowdcosy=<y of row direction cosine>
#  rowdcosy=<z of row direction cosine>
#  coldcosx=<x of col direction cosine>
#  coldcosy=<y of col direction cosine>
#  coldcosz=<z of col direction cosine>
#  rowspc=<spacing between columns>
#  colspc=<spacing between rows>
#  ztol=<allowable z-distance from contour to image plane>
#
# On the input socket, the format of the contours is the following:
# BEGIN CONTOUR             - marks the beginning of a contour
# x1,y1,z                   - first point in contour
# x2,y2,y2                  - next point
# ...                       - and so on
# xn,yn,zn                  - last point
# END CONTOUR
# ...                       - repeat for as many contours as you have
#
use strict;
use Posda::FlipRotate;
use Math::Polygon;

my($in, $out, $status, $rows, $cols, $ulx, $uly, $ulz,
  $rowdcosx, $rowdcosy, $rowdcosz,
  $coldcosx, $coldcosy, $coldcosz,
  $rowspc, $colspc,
  $ztol
);
for my $i (@ARGV){
  unless($i =~ /^([^=]+)=([^=]+)$/) {
    die "$0: can't parse parameter $i";
  }
  my $key = $1;
  my $value = $2;
  if($key eq "in") { $in = $value }
  elsif ($key eq "out") { $out = $value }
  elsif ($key eq "status") { $status = $value }
  elsif ($key eq "rows") { $rows = $value }
  elsif ($key eq "cols") { $cols = $value }
  elsif ($key eq "ulx") { $ulx = $value }
  elsif ($key eq "uly") { $uly = $value }
  elsif ($key eq "ulz") { $ulz = $value }
  elsif ($key eq "rowdcosx") { $rowdcosx = $value }
  elsif ($key eq "rowdcosy") { $rowdcosy = $value }
  elsif ($key eq "rowdcosz") { $rowdcosz = $value }
  elsif ($key eq "coldcosx") { $coldcosx = $value }
  elsif ($key eq "coldcosy") { $coldcosy = $value }
  elsif ($key eq "coldcosz") { $coldcosz = $value }
  elsif ($key eq "rowspc") { $rowspc = $value }
  elsif ($key eq "colspc") { $colspc = $value }
  elsif ($key eq "ztol") { $ztol = $value }
  else { die "$0: unknown parameter: $key" }
}
unless(defined $in) { die "$0: in is not defined" }
unless(defined $out) { die "$0: out is not defined" }
unless(defined $rows) { die "$0: cols is not defined" }
unless(defined $ulx) { die "$0: ulx is not defined" }
unless(defined $uly) { die "$0: uly is not defined" }
unless(defined $ulz) { die "$0: ulz is not defined" }
unless(defined $rowdcosx) { die "$0: rowdcosx is not defined" }
unless(defined $rowdcosy) { die "$0: rowdcosy is not defined" }
unless(defined $rowdcosz) { die "$0: rowdcosz is not defined" }
unless(defined $coldcosx) { die "$0: coldcosx is not defined" }
unless(defined $coldcosy) { die "$0: coldcosy is not defined" }
unless(defined $coldcosz) { die "$0: coldcosz is not defined" }
unless(defined $rowspc) { die "$0: rowspc is not defined" }
unless(defined $colspc) { die "$0: colspc is not defined" }
unless(defined $ztol) { die "$0: ztol is not defined" }
open(INPUT, "<&", $in) or die "$0: Can't open in = $in ($!)";
open(OUTPUT, ">&", $out) or die "$0: Can't open out = $out ($!)";

my @contours;
my $contour = [];
my $mode = "scan";
while(my $line = <INPUT>){
  chomp $line;
  while($mode eq "scan" && $line ne "BEGIN CONTOUR") { next }
  if($mode eq "scan"){
    $mode = "in contour";
    next;
  }
  if($line eq "END CONTOUR"){
    if($#{$contour} >= 0){
      my $first = $contour->[0];
      my $last = $contour->[$#{$contour}];
       unless(
        $first->[0] eq $last->[0] &&
        $first->[1] eq $last->[1] &&
        $first->[2] eq $last->[2]
      ){
        push(@{$contour}, $first);
      }
    }
    push @contours, $contour;
    $contour = [];
    $mode = "scan";
    next;
  }
  my @numbers = split(/\\/, $line);
  my $num_numbers = @numbers;
  unless($num_numbers % 3 == 0){
    die "$0: Number of numbers ($num_numbers) isn't divisible by 3";
  }
  my $num_points = $num_numbers / 3;
  for my $i (0 .. $num_points - 1){
    my $point = [$numbers[$i*3], $numbers[($i*3)+ 1], $numbers[($i*3) + 2]];
    push(@{$contour}, $point);
  }
}
#my $num_contours = @contours;
#for my $i (0 .. $#contours){
#  my $count = @{$contours[$i]};
#}
unless($mode eq "scan") { die "$0: end of input in scan mode" }
my @contours_2d;
my $ipp = [$ulx, $uly, $ulz];
my $iop = [$rowdcosx, $rowdcosy, $rowdcosz, $coldcosx, $coldcosy, $coldcosz];
my $pix_sp = [$rowspc, $colspc];
for my $i (0 .. $#contours){
  my $contour_2d = [];
  for my $j (@{$contours[$i]}){
    my $p = Posda::FlipRotate::ToPixCoords(
      $iop, $ipp, $rows, $cols, $pix_sp, $j);
    my($x, $y, $z) = @$p;
    if($z > $ztol) { 
      die "$0: point ($j->[0], $j->[1], $j->[2]) off plane by $z (> $ztol)"
    }
    push(@{$contour_2d}, [$x, $y]);
  }
  my $s = $contour_2d->[0];
  my $e = $contour_2d->[$#{$contour_2d}];
  if(
    sqrt(
      (($s->[0] - $e->[0]) **2) +
      (($s->[0] - $e->[0]) **2)
    ) < ((($rowspc + $colspc) / 2) / 2)
  ){
    $contour_2d->[$#{$contour_2d}] = $contour_2d->[0];
  } else {
    push(@{$contour_2d}, $contour_2d->[0]);
  }
  push(@contours_2d, $contour_2d);
}
my @segments;
for my $c (@contours_2d){
  for my $i (0 .. $#{$c}- 1){
    my $seg = [$c->[$i], $c->[$i+1]];
    push(@segments, $seg);
  }
  push(@segments, [$c->[$#{$c}], $c->[0]]);
}
@segments = sort { $a->[0]->[0] <=> $b->[0]->[0] } @segments;
my $bit_count = 0;
my $byte = 0;
my $mask = 0x01;
for my $i (0 .. $rows - 1){
  my @cross_segs;
  for my $s (@segments){
    if(
      ($s->[0]->[1] <= $i && $s->[1]->[1] > $i) ||
      ($s->[1]->[1] <= $i && $s->[0]->[1] > $i)
    ){ push @cross_segs, $s }
  }
  my $num_cross = @cross_segs;
  if($num_cross & 1) {
    print STDERR "$0: Odd number crossings (Not closed planar?)\n";
  }
  my $next_crossing;
  if($#cross_segs >= 0){
    $next_crossing = calc_crossing(shift(@cross_segs), $i);
  }
  my $crossing_polarity = 0;
  for my $j (0 .. $cols - 1){
    if(defined($next_crossing) && $j > $next_crossing) {
      $crossing_polarity ^= 1;
      if($#cross_segs >= 0){
        $next_crossing = calc_crossing(shift(@cross_segs), $i);
      } else {
        $next_crossing = undef;
      }
    }
    if($crossing_polarity) {
      $byte |= $mask;
    }
    $mask *= 2;
    $bit_count += 1;
    if($bit_count == 8){
      no warnings;
      print OUTPUT pack("c", $byte);
      $byte = 0;
      $bit_count = 0;
      $mask = 0x01;
    }
  }
}
if($bit_count > 0){
  no warnings;
  print OUTPUT pack("c", $byte);
}
if(defined $status){
  open(STATUS, ">&", $status) or die "$0: Can't open status = $status ($!)";
  print STATUS "OK\n";
  close STATUS;
}
sub calc_crossing{
  my($seg, $y) = @_;
  my $start;
  my $num;
  my $denom;
  my $dist;
  $start = $seg->[0]->[0];
  $dist = $seg->[1]->[0] - $seg->[0]->[0];
  $num = $y - $seg->[0]->[1];
  $denom = $seg->[1]->[1] - $seg->[0]->[1];
  return $start + ($dist * ($num/$denom));
}
